### PR TITLE
add utilities for migrating services to ASF

### PR DESCRIPTION
**Pull request by thrau**
This PR adds code utilities and mechanisms to migrate services to the new ASF as described in #5222 

* adds the `AsfWithFallbackListener` and `AsfWithPersistingFallbackListener` as described in #5222 
* SNS example can be found in the [asf-migration-sns](https://github.com/localstack/localstack/tree/asf-migration-sns) branch in da433bdd8b571e43e67ca214438d4a32f41e85ac. however, this would currently break the pro SNS provider that needs to be updated.

### CODE REVIEW

File: tests/unit/aws/test_proxy.py [hunk 1]
```diff
@@ -35,3 +37,55 @@ def list_queues(self, context, request):
             assert result["QueueUrls"] == [
                 "http://localhost:4566/000000000000/foo-queue",
             ]
+
+
+class TestAsfWithFallbackListener:
+    def test_request_response(self):
+        provider_calls = []
+        fallback_calls = []
+
```
Comment (calvernaz) [inline tests/unit/aws/test_proxy.py:46]: I can see this setup being widely reused, the pair `Provider`/`Fallback` would be interesting to have fixtures to pull them into different tests and not pollute the setup.

File: localstack/aws/skeleton.py [hunk 2]
```diff
@@ -151,26 +149,52 @@ def invoke(self, context: RequestContext) -> HttpResponse:
                     operation.name,
                 )
                 raise NotImplementedError
-            handler = self.dispatch_table[operation.name]
 
-            # Call the appropriate handler
-            result = handler(context, instance) or {}
-            # Serialize result dict to an HTTPResponse and return it
-            return serializer.serialize_to_response(result, operation)
+            return self.dispatch_request(context, instance)
```
Comment (calvernaz) [inline localstack/aws/skeleton.py:153]: potential nic, do you think `skeleton` could be called `service dispatcher` or `request dispatcher`, the concepts within are all screaming "service" and "dispatch", just a thought tho.

**Review by calvernaz — APPROVED** [review]
Is looking good, the abstractions are very clear and the code looks good! 👍 

small nics, but overall pretty good

File: localstack/aws/skeleton.py (hunk 2 above)
Comment (thrau) [reply localstack/aws/skeleton.py:153]: yep agreed. Skeleton just came from the classic textbook RPC architecture (i.e., the server stub) :-) actually the Skeleton will be replaced soon-ish with the HandlerChain framework (currently underway in the asf branch).

File: tests/unit/aws/test_proxy.py (hunk 1 above)
Comment (thrau) [reply tests/unit/aws/test_proxy.py:46]: yep good point. it would probably be good to use mocks in this case that already record calls internally.

**Pull request merged**